infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/fsm/sink/RemoteSourceLeadershipManager.java

```diff
@@ -0,0 +1,256 @@
+
+                    if (leadershipResponse.getIsLeader()) {
+                        log.info("Received Leadership Response :: leader for remote cluster, node={}", leadershipResponse.getNodeId());
+                        leader = leadershipResponse.getNodeId();
+                        setRemoteLeaderNodeId(leader);
+
+                        // Remove all CF, based on the assumption that one leader response is the expectation.
+                        pendingLeadershipQueries.clear();
+
+                        // A new leader has been found,
+                        input(new LogReplicationSinkEvent(LogReplicationSinkEvent.LogReplicationSinkEventType.REMOTE_LEADER_FOUND, leader));
+                        router.getSessionToLeaderConnectionFuture().get(session).complete(null);
+                        log.debug("Exit :: leadership verification");
+                        return;
+                    } else {
+                        log.debug("Received Leadership Response :: node {} is not the leader", leadershipResponse.getNodeId());
+
+                        // Remove CF for completed request
+                        pendingLeadershipQueries.remove(leadershipResponse.getNodeId());
+                    }
+                }
+
+                // No remote leader was found, retry leadership
+                input(new LogReplicationSinkEvent(LogReplicationSinkEvent.LogReplicationSinkEventType.REMOTE_LEADER_NOT_FOUND, leader));
+
+            } catch (Exception ex) {
+                log.warn("Exception caught while verifying remote leader.", ex);
+                input(new LogReplicationSinkEvent(LogReplicationSinkEvent.LogReplicationSinkEventType.REMOTE_LEADER_NOT_FOUND, leader));
+            }
+        } else {
+            log.info("Remote Leader already present {}. Skip leader verification.", getRemoteLeaderNodeId().get());
+        }
+
+        log.debug("Exit :: leadership verification");
+    }
+
+    private void invokeReverseReplication() {
+        CorfuMessage.ResponsePayloadMsg payload =
+                CorfuMessage.ResponsePayloadMsg.newBuilder()
+                        .setLrSubscribeMsg(LogReplication.SubscribeToReplicationMsg
+                                .newBuilder()
+                                .setSinkLeaderNodeId(localNodeId)
+                                .build())
+                        .build();
+
+        CorfuMessage.HeaderMsg header = CorfuMessage.HeaderMsg.newBuilder()
+                .setSession(session)
+                .setRequestId(router.getSessionToRequestIdCounter().get(session).getAndIncrement())
+                .setClusterId(getUuidMsg(UUID.fromString(session.getSourceClusterId())))
+                .setVersion(getDefaultProtocolVersionMsg())
+                .setIgnoreClusterId(true)
+                .setIgnoreEpoch(true)
+                .build();
+
+
+        log.info("Sending the subscribe msg {} for session {}", payload, session);
```

Comment:
You can change the log line to reflect reverse replication instead of subscribe terminology.